### PR TITLE
new package net-misc/ptpd

### PR DESCRIPTION
**Pull request by j123b567**
PTPd is implementation of Precise Time Protocol (IEEE 1588)

see also https://bugs.gentoo.org/show_bug.cgi?id=492508

### CODE REVIEW

**Comment by mgorny** [comment]
Are you going to maintain this package? If so, please fill in `<maintainer/>` in metadata.xml, and add `<herd>proxy-maint</herd>` above it.

**Comment by j123b567** [comment]
Yes, I will maintain this package.

File: net-misc/ptpd/metadata.xml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+  <herd>proxy-maint</herd>
+  <maintainer>
+    <email>jan.breuer@jaybee.cz</email>
+	<name>Jan Breuer</name>
```
Comment (mgorny) [inline net-misc/ptpd/metadata.xml]: Mis-indented :).

File: net-misc/ptpd/metadata.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+  <herd>proxy-maint</herd>
```
Comment (mgorny) [inline net-misc/ptpd/metadata.xml]: Sorry, I was wrong. It's supposed to be `proxy-maintainers`.

File: net-misc/ptpd/files/ptpd-2.3.0-ntpdc.patch [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ 
+ 	if(!hostLookup(options->hostAddress, &control->serverAddress)) {
+                 control->serverAddress = 0;
+@@ -236,8 +236,7 @@
+ 	int	maclen;
+ 	static char *key;
+ 
+-	key=malloc(21);
+-	memset (key, 0, sizeof(key));
++	key=calloc(21,sizeof(char));
```
Comment (mgorny) [inline net-misc/ptpd/files/ptpd-2.3.0-ntpdc.patch:19]: Just a fun fact about `sizeof`: _When applied to an operand that has type char, unsigned char, or signed char,
(or a qualified version thereof) the result is 1._ (C99)

File: net-misc/ptpd/files/ptpd-2.3.0-ntpdc.patch [hunk 4]
```diff
@@ -0,0 +1,22 @@
+diff -urN ptpd-2.3.0.old/src/dep/ntpengine/ntpdcontrol.c ptpd-2.3.0.new/src/dep/ntpengine/ntpdcontrol.c
```
Comment (mgorny) [inline net-misc/ptpd/files/ptpd-2.3.0-ntpdc.patch:1]: A generic question here: did you submit this patch upstream? Since it looks like a genuine bugfix :)

Comment (mgorny) [inline net-misc/ptpd/files/ptpd-2.3.0-ntpdc.patch:1]: Sorry, I now see it's no longer needed in 2.3.1*, so ok :).

File: net-misc/ptpd/files/ptpd-2.3.0-statistics-clear.patch [hunk 5]
```diff
@@ -0,0 +1,12 @@
+diff -urN ptpd-2.3.0.old/src/dep/statistics.c ptpd-2.3.0/src/dep/statistics.c
+--- ptpd-2.3.0.old/src/dep/statistics.c	2013-11-22 17:17:58.000000000 +0100
++++ ptpd-2.3.0/src/dep/statistics.c	2015-06-07 10:29:45.697110725 +0200
+@@ -528,7 +528,7 @@
+ void
+ clearPtpEngineSlaveStats(PtpEngineSlaveStats* stats)
+ {
+-	memset(stats, 0, sizeof(stats));
++	memset(stats, 0, sizeof(PtpEngineSlaveStats));
```
Comment (mgorny) [inline net-misc/ptpd/files/ptpd-2.3.0-statistics-clear.patch]: I'd use `*stats` but that's just me :).

File: net-misc/ptpd/files/ptpd.rc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+description="Precise Time Protocol daemon"
+
+
+depend() {
+        use net dns logger
+}
+
+checkconfig() {
+	ebegin "Checking ${SVCNAME} configuration"
+        msgout="$($command -k $command_args 2>&1)"
```
Comment (mgorny) [inline net-misc/ptpd/files/ptpd.rc]: Looks like indent going out of hand. Please make sure to indent using tabs, not spaces.

File: net-misc/ptpd/files/ptpd.rc [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	eend $RETVAL
+}
+
+start() {
+        checkconfig || return 1
+
+	ebegin "Starting ${SVCNAME}"
+	start-stop-daemon --start --exec ${command} \
+		--pidfile ${pidfile} \
+		-- ${command_args} 1>/dev/null 2>&1
```
Comment (mgorny) [inline net-misc/ptpd/files/ptpd.rc]: Why are you killing all output?

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+EAPI=5
+
+inherit autotools eutils flag-o-matic systemd
+
+DESCRIPTION="Precision Time Protocol daemon"
+HOMEPAGE="http://ptpd.sf.net"
+
+if [[ ${PV} == "9999" ]]; then
+	ESVN_REPO_URI="http://svn.code.sf.net/p/ptpd/code/trunk"
+	inherit subversion
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: You should avoid calling `inherit` more than once. Usually people put the VCS eclass in a variable and do like `inherit autotools eutils ... ${VCS_ECLASS}`

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+# $Header: $
+
+EAPI=5
+
+inherit autotools eutils flag-o-matic systemd
+
+DESCRIPTION="Precision Time Protocol daemon"
+HOMEPAGE="http://ptpd.sf.net"
+
+if [[ ${PV} == "9999" ]]; then
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: While I don't mind strongly having live code there, it feels a bit weird to not have -9999 ebuild that could use it :).

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	ESVN_REPO_URI="http://svn.code.sf.net/p/ptpd/code/trunk"
+	inherit subversion
+	SRC_URI=""
+	KEYWORDS=""
+else
+	if [[ ${PV} == *_rc* ]]; then
+		MY_PV=${PV/_rc*/}
+		MY_P=${P/_rc/-rc}
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: I don't think doing those substitutions on non-rc P/PV would change anything. So why not remove the conditional and just do them unconditionally?

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	fi
+
+	SRC_URI="mirror://sourceforge/ptpd/${MY_PV}/${MY_P}.tar.gz"
+	KEYWORDS="~amd64 ~x86 ~arm"
+fi
+
+LICENSE="BSD"
+SLOT="0"
+IUSE="snmp +statistics ntp experimental debug +daemon"
+COMMON_DEPEND=" snmp? ( net-analyzer/net-snmp )
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: Please don't space-align. Just tab indent, one indent for deps.

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                ntp? ( net-misc/ntp )
+                net-libs/libpcap"
+RDEPEND="${COMMON_DEPEND}"
+DEPEND="${COMMON_DEPEND}"
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: Please don't do COMMON_DEPEND when there's no non-common deps. Just regular RDEPEND=foo, DEPEND=${RDEPEND}.

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+S="${WORKDIR}/${MY_P}"
+
+src_unpack() {
+	if [[ ${PV} == "9999" ]] ; then
+		subversion_src_unpack
+	else
+		unpack ${A}
+		cd "${S}"
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: epatch belongs in src_prepare(), then cd is unneeded.

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+		epatch "${FILESDIR}/${P}-statistics-clear.patch"
+		epatch "${FILESDIR}/${P}-ntpdc.patch"
+	fi
+}
+
+src_prepare() { eautoreconf; }
+
+src_configure() {
+	append-flags -fno-strict-aliasing
+	econf \
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild:33]: Personal suggestion:

```
local myconf=(
    $(use_enable snmp)
    # this way you can have inline comments!
    $(use_enable experimental experimental-options)
)
econf "${myconf[@]}"
```

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+		$(use_enable snmp) \
+		$(use_enable experimental experimental-options) \
+		$(use_enable statistics) \
+		$(use_enable ntp ntpdc) \
+		$(use_enable debug runtime-debug) \
+		$(use_enable daemon)
+}
+
+src_install() {
+	emake install DESTDIR="${D}" || die "install failed"
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: emake calls `die` implicitly in EAPI 4 and newer.

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+	insinto /etc
+	newins "${S}/src/ptpd2.conf.minimal" ptpd2.conf
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: Explicit ${S} is unnecessary here (cwd is ${S}).

File: net-misc/ptpd/ptpd-2.3.0.ebuild [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+	dodoc "${S}/src/ptpd2.conf.default-full"
+	dodoc "${S}/src/ptpd2.conf.minimal"
+
+	newinitd "${FILESDIR}/ptpd.rc" ptpd2
+	newconfd "${FILESDIR}/ptpd.confd" ptpd2
+
+	systemd_newunit "${FILESDIR}/ptpd.service" ptpd2.service
```
Comment (mgorny) [inline net-misc/ptpd/ptpd-2.3.0.ebuild]: Why not name the file `ptpd2.service` and avoid newunit?

**Comment by j123b567** [comment]
I have tried to fix all these issues (see extended commit description). I just did not refactor `econf` to use `local myconf = ...`.

**Comment by mgorny** [comment]
Ok, I'm going to try building it and then commit :).

**Comment by mgorny** [comment]
Few extra notes for the future (I'll fix them myself):
- please sort dependencies, USE flags, keywords lexically. It makes easier for others (e.g. arch teams) to find them.
- please use `:=` for rdepends that the package links to, to get it automatically rebuilt when soname changes,
- run `repoman full` and fix all the issues :).

**Comment by mgorny** [comment]
```
+*ptpd-2.3.1_rc4 (12 Jun 2015)
+*ptpd-2.3.0 (12 Jun 2015)
+
+  12 Jun 2015; Michał Górny <mgorny@gentoo.org> +files/ptpd-2.3.0-ntpdc.patch,
+  +files/ptpd-2.3.0-statistics-clear.patch,
+  +files/ptpd-2.3.1_rc4-debug-display.patch, +files/ptpd2.confd,
+  +files/ptpd2.rc, +files/ptpd2.service, +metadata.xml, +ptpd-2.3.0.ebuild,
+  +ptpd-2.3.1_rc4.ebuild:
+  New ebuild from Jan Breuer: PTPd -- implementation of Precise Time Protocol
+  (IEEE 1588). https://github.com/gentoo/gentoo-portage-rsync-mirror/pull/144.
```

**Pull request closed without merging**